### DISCUSSION TITLE
Issues for supporting QMK 0.14.x

### DISCUSSION THREAD

**Discussion by koron** [Yowkees/keyball#68] [General]
Keyball46のファームウェアをQMK 0.14.xに対応させるための課題や仮定を記録として残すためのスレッドです。

---

cf. [現状のdiff](https://github.com/koron/qmk_firmware/compare/kaoriya...koron:keyball46-catchup-qmk0.14)

**Comment by koron** [comment]
まず現状わかってる範囲では以下の問題が判明しました。

1. シリアル通信の方式が変わっており、トラックボールのデータを送るためにフォークしていたコードがコンパイルできなくなっていた
2. OLEDを有効化するためのフラグ名が変わっており、OLEDが無効化されていた
3. ファームウェアサイズが大きくなり、制限である28KBを約400バイト超えてしまった
    特にVIA, RGBLIGHT, OLEDのフルセットを有効化できていない

このうち1と2は既に対応の目途がたっていますが、3はまだです。

**Reply by Yowkees** [reply]
1については把握しておりましたが、2の問題と、QMKバージョンアップに付随して発生した3については把握できておりませんでした。
ご指摘とご対応、誠にありがとうございます！

**Comment by koron** [comment]
解説します。

> シリアル通信の方式が変わって

QMKのsplitにおけるシリアル通信は疑似的なメモリ共有になっています。
つまりprimaryがあるメモリ領域に書いた内容がそのままsecondary側に見え、
逆にsecondaryがあるメモリ領域に書いた内容がそのままprimaryに見えるという感じです。

0.13.xまではこのメモリ領域=構造体が1つだけだったので
Keyball46ではその構造体にトラックボール用のデータを無理やり追加して
secondary boardのトラックボールを動かせるようにしてました。

しかし0.14.xではこのメモリ領域=トランザクションを複数持てるようになり、
さらに各キーボードファームで追加拡張できるように変更されました。
この変更内容そのものはエンコーダーやRGB MATRIXといった機能でもただしく利用可能になったことから
極めて正しいと評価できます。
しかし無理やりトラックボールのデータを追加していたKeyball46は
これによりコンパイルできなくなりました。

Keyball46における対策としては、
トランザクションの追加拡張機能を用いてトラックボールデータをやりとりするように変更します。

(これはもうほぼほぼできている)

**Comment by koron** [comment]
> OLEDを有効化するためのフラグ名が変わって

これは非常に簡単で
以前は `OLED_DRIVER_ENABLE` だったものが `OLED_ENABLE` に変更になりました。

Keyball46のファームでもこれに追従する形で変更します。

**Comment by koron** [comment]
> ファームウェアサイズが大きくなり

具体的にどこが大きくなったのかはわかりませんが
Keyball46側ではなくQMK側の各種変更の影響だと推測しています。

その影響により VIA, RGB LIGHT, OLED の3つの機能を同時に有効化すると
Pro Microの制限である28KBを約400Bを超えてしまいます。

ファーム内の不変なデータの内訳を見てみると…

* 光学センサーのファーム - 4KB
* OLEDのフォントデータ - 1.3KB
* デフォルトのキーマップ - 384B

あたりが目立って大きなところです。
なおこれらのデータの圧縮は過去に検討してて効果がないことがわかっています。

かなり煮詰まっていますが、
あとは各機能ごとに細かく無効化できるものがないか
精査して切り詰めるしかないかなという感じはしてます。

**Reply by Yowkees** [reply]
ご解説、誠にありがとうございます。よく理解できました。

QMK側の左右通信の拡張機能が拡張され、おそらく別機能も増えたことから、サイズが大きくなってしまいました。
私の方でも何か無効化できる機能がないか調べてみたいと思います。

**Comment by koron** [comment]
いろいろ考えてみたけど
VIAなファームはいったんOLEDを無効化して
代わりにRGBLIGHTのフル機能を有効化して
リリースするのが良いんじゃないかと思えてきました。

0.13.xの時でもサイズの関係でRGBLIGHTの一部(というか大半)の機能を無効化していました。
しかしVIAにはRGBLIGHTの制御機能があるのでせっかくの機能が持ち腐れという状況だったんです。

VIA対応ファームはファーム初心者にとっては大事な入口です。
サイズの関係でどちらかを絶対に落とさなければならないのであれば
RGBLIGHTのほうがVIAからいじれた方が
うれしいのではないかとの結論に至りました。

代わりにVIAを含まない default ファームは
独自にファームを書く人のテンプレートともなりますし
OLED+フルRGBLIGHTの構成にするとその役割分担的にも意味が生じます。

まずはこのラインでPR作ります。

**Reply by koron** [reply]
結果、ギリギリ。0.15.x ではコレもNGになりそうw

```
Checking file size of keyball_rev1_ball_via.hex                                                     [WARNINGS]

 * The firmware size is approaching the maximum - 28028/28672 (97%, 644 bytes free)
 ```

**Reply by koron** [reply]
defaultでフルRGBLIGHT+OLEDがだめだったので
一部のRGBLIGHTエフェクトだけをこんなセレクションで有効化した結果…

```
#    define RGBLIGHT_EFFECT_BREATHING
#    define RGBLIGHT_EFFECT_RAINBOW_MOOD
//#    define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#    define RGBLIGHT_EFFECT_SNAKE
//#    define RGBLIGHT_EFFECT_KNIGHT
//#    define RGBLIGHT_EFFECT_CHRISTMAS
//#    define RGBLIGHT_EFFECT_STATIC_GRADIENT
#    define RGBLIGHT_EFFECT_RGB_TEST
#    define RGBLIGHT_EFFECT_ALTERNATING
//#    define RGBLIGHT_EFFECT_TWINKLE
```

```
Checking file size of keyball_rev1_ball_default.hex                                                 [WARNINGS]

 * The firmware size is approaching the maximum - 28672/28672 (100%, 0 bytes free)
```

1バイトすらも猶予が無いファームができあがりましたwww

**Reply by koron** [reply]
Keyball46はLEDの数が左右合わせて14個と多くはないので
「映える」エフェクトは多くなさそう。

全エフェクトをざっくりと評価するとこんな感じになった。

エフェクト                     | 映え | 備考
-------------------------------|------|----------------------------
RGBLIGHT_EFFECT_BREATHING      | HIGH |全体が一斉に変わるので映える
RGBLIGHT_EFFECT_RAINBOW_MOOD   | HIGH |全体が一斉に変わるので映える
RGBLIGHT_EFFECT_RAINBOW_SWIRL  | LOW  |少ないLEDにグラデーションが合わない
RGBLIGHT_EFFECT_SNAKE          | LOW  |少ないLEDで動きが分かりにくい
RGBLIGHT_EFFECT_KNIGHT         | LOW  |少ないLEDで動きが分かりにくい
RGBLIGHT_EFFECT_CHRISTMAS      | MID  |わかりやすいが、面白みに欠ける
RGBLIGHT_EFFECT_STATIC_GRADIENT| LOW  |少ないLEDにグラデーションが合わない
RGBLIGHT_EFFECT_RGB_TEST       | HIGH |オーソドックスなテスト・デモに向く
RGBLIGHT_EFFECT_ALTERNATING    | HIGH |交互に光って…テスト用。RGB TESTとどっちか片方でも良い
RGBLIGHT_EFFECT_TWINKLE        | MID  |綺麗だが光り方がまばら。もっとLEDが多ければ…

この映え優先度で有効化できないかやってみる。

**Reply by koron** [reply]
やってみたらHIGHにしたものが入った。MIDの2つはどちらかも、片方入れただけであふれた。

なのでdefaultで有効化するRGBLIGHTエフェクトは上記の通りとします。

あとVIAではtwinkleが入りませんでした。
こいつだけで910バイトもあるよ…

**Comment by koron** [comment]
とりあえず上記の内容でPR作り https://github.com/Yowkees/keyball/pull/69

ビルドしたファームをアップしておきました https://github.com/Yowkees/keyball/discussions/70

**Reply by Yowkees** [reply]
多大なるご尽力をいただき、本当にありがとうございます！！
json等のメンテナンスまで、、頭が上がりません(ToT)
またメモリ不足に対して素晴らしい解決策をお考えいただき、本当にありがとうございましたm(_ _)m

Keyball46購入者様は、
1. まずはVIAファームを入れてもらい、キーマップを気軽に書き換えて使用していただく
2. キーマップがある程度固まり、かつOLEDをカスタマイズしたい or ファームを改造したいとなったらdefaultファームをイジイジしてもらう
という流れと理解しました。

ファームウェアについて質問があった際には対応したいと思います。
PRをマージさせていただきます。

改めまして、本当に多くの注力をいただきまして、誠にありがとうございました！
